pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -166,15 +178,25 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
 	// We will wait for all predecessors to be Running and Ready prior to attempting a deletion.
 	// We will terminate Pods in a monotonically decreasing order over [len(pods),set.Spec.Replicas).
 	// Note that we do not resurrect Pods in this interval.
-	if unhealthy > 0 {
+	if unhealthy > 0 && monotonic {
 		glog.V(2).Infof("StatefulSet %s is waiting on %d Pods", set.Name, unhealthy)
 		return nil
 	}
-	if target := len(condemned) - 1; target >= 0 {
+	for target := len(condemned) - 1; target >= 0; target-- {
 		glog.V(2).Infof("StatefulSet %s terminating Pod %s", set.Name, condemned[target])
-		return ssc.podControl.DeleteStatefulPod(set, condemned[target])
+		if err := ssc.podControl.DeleteStatefulPod(set, condemned[target]); err != nil {
+			return err
+		}
+		if monotonic {
+			return nil
+		}
 	}
 	return nil
 }
 
 var _ StatefulSetControlInterface = &defaultStatefulSetControl{}
+
+// setAllowsBurst is true if the alpha burst annotation is set.
+func setAllowsBurst(set *apps.StatefulSet) bool {
```

Comment:
Move the helper to utils. If you keep the annotation instead of going with a field you're going to have to validate it here. You should report an error to  runtime.ReportError if annotations value is not boolean.